From 2be9f8c54a2edb25a83079deef1ade69e8454414 Mon Sep 17 00:00:00 2001
From: De Huo <De.Huo@windriver.com>
Date: Thu, 1 Aug 2019 19:12:37 +0800
Subject: [PATCH] Add rule accounts_passwords_pam_tally2_interval for
 WRLinux1019 profile

Add rule accounts_passwords_pam_tally2_interval test case for
 WRLinux1019 product, and add this rule in WRLinux1019 profile. 

Upstream-Status: Pending

Signed-off-by: De Huo <De.Huo@windriver.com>
---
 .../ansible/shared.yml                             | 81 ++++++++++++++++++++++
 .../bash/wrlinux1019.sh                            | 14 ++++
 .../oval/wrlinux1019.xml                           | 31 +++++++++
 .../rule.yml                                       | 44 ++++++++++++
 ...counts_passwords_pam_faillock_fail_interval.var | 19 +++++
 .../profiles/draft_stig_wrlinux_disa.profile       |  1 +
 6 files changed, 190 insertions(+)
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/ansible/shared.yml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/bash/wrlinux1019.sh
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/oval/wrlinux1019.xml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/rule.yml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pwquality/accounts_passwords_pam_faillock_fail_interval.var

diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/ansible/shared.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/ansible/shared.yml
new file mode 100644
index 0000000..d2b08c0
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/ansible/shared.yml
@@ -0,0 +1,81 @@
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_rhv
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+- (xccdf-var var_accounts_passwords_pam_faillock_deny)
+
+- name: Add auth pam_faillock preauth deny before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: required
+    new_module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: before
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock preauth
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: required
+    module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add auth pam_faillock authfail deny after pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: '[default=die]'
+    new_module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: after
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock authfail
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    new_type: auth
+    control: '[default=die]'
+    module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add account pam_faillock before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: account
+    control: required
+    module_path: pam_unix.so
+    new_type: account
+    new_control: required
+    new_module_path: pam_faillock.so
+    state: before
+  loop:
+    - system-auth
+    - password-auth
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/bash/wrlinux1019.sh b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/bash/wrlinux1019.sh
new file mode 100644
index 0000000..e0dc6a0
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/bash/wrlinux1019.sh
@@ -0,0 +1,14 @@
+# platform = multi_platform_wrlinux
+
+. /usr/share/scap-security-guide/remediation_functions
+populate var_accounts_passwords_pam_faillock_fail_interval
+
+if grep -q '^[^#].*pam_tally2.so' /etc/pam.d/common-auth; then
+  if grep -q 'lock_time=' /etc/pam.d/common-auth; then
+	  sed -i "s/lock_time=[0-9]*/lock_time=$var_accounts_passwords_pam_faillock_fail_interval/" /etc/pam.d/common-auth
+  else
+	  sed -i "/auth.*pam_tally2.so/s/$/ lock_time=$var_accounts_passwords_pam_faillock_fail_interval/" /etc/pam.d/common-auth
+  fi
+else
+  echo "auth  required pam_tally2.so lock_time=$var_accounts_passwords_pam_faillock_fail_interval" >>/etc/pam.d/common-auth
+fi
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/oval/wrlinux1019.xml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/oval/wrlinux1019.xml
new file mode 100644
index 0000000..1a75370
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/oval/wrlinux1019.xml
@@ -0,0 +1,31 @@
+<def-group>
+  <definition class="compliance" id="accounts_passwords_pam_tally2_interval" version="2">
+    <metadata>
+      <title>Always deny for specified seconds after failed attempt</title>
+      <affected family="unix">
+        <platform>multi_platform_wrlinux</platform>
+      </affected>
+      <description>The number of allowed failed logins should be set correctly.</description>
+    </metadata>
+    <criteria>
+      <criterion comment="check pam_tally2 option lock_time" test_ref="test_accounts_passwords_pam_tally2_lock_time" />
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test check="all" check_existence="all_exist" comment="check maximum lock_time allowed in /etc/pam.d/common-auth" id="test_accounts_passwords_pam_tally2_lock_time" version="2">
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_lock_time" />
+    <ind:state state_ref="state_accounts_passwords_pam_tally2_lock_time" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_lock_time" version="2">
+    <ind:filepath>/etc/pam.d/common-auth</ind:filepath>
+    <ind:pattern operation="pattern match">^\s*auth\s+(?:(?:required))\s+pam_tally2\.so\s+.*\block_time=([0-9]*).*$</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_state id="state_accounts_passwords_pam_tally2_lock_time" version="2">
+    <ind:subexpression datatype="int" operation="greater than or equal" var_ref="var_accounts_passwords_pam_faillock_fail_interval" />
+  </ind:textfilecontent54_state>
+
+  <external_variable comment="maximum lock_time allowed for pam_tally2" datatype="int" id="var_accounts_passwords_pam_faillock_fail_interval" version="2" />
+</def-group>
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/rule.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/rule.yml
new file mode 100644
index 0000000..a53bdb1
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_interval/rule.yml
@@ -0,0 +1,44 @@
+documentation_complete: true
+
+prodtype: wrlinux1019
+
+title: 'Set Interval For Counting Failed Password Attempts'
+
+description: |-
+    Utilizing <tt>pam_tally2.so</tt>, the <tt>lock_time</tt> directive
+    configures the system to lock out an accounts after a number of incorrect login
+    attempts within a specified time period.. Modify the content of both
+    <tt>/etc/pam.d/common-auth</tt> and <tt>/etc/pam.d/common-account</tt> as follows:
+    <br /><br />
+    <ul>
+    <li>Add the following line immediately <tt>before</tt> the <tt>pam_unix.so</tt> statement in the <tt>/etc/pam.d/common-auth</tt>:
+    <pre>auth required pam_tally2.so deny=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_deny" use="legacy"/> lock_time=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_fail_interval" use="legacy"/> even_deny_root root_unlock_time=60 unlock_time=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_unlock_time" use="legacy"/> </pre></li>
+    <li> add the following line in the <code>/etc/pam.d/common-account</code>:
+    <pre>account required pam_tally2.so </pre></li>
+    </ul>
+
+rationale: |-
+    By limiting the number of failed logon attempts the risk of unauthorized system
+    access via user password guessing, otherwise known as brute-forcing, is reduced.
+    Limits are imposed by locking the account.
+
+severity: medium
+
+identifiers:
+    cce@rhel7: 27297-1
+
+references:
+    nist: AC-7(b)
+    srg: SRG-OS-000329-GPOS-00128,SRG-OS-000021-GPOS-00005
+    stigid@rhel7: "010371"
+
+ocil_clause: 'that is not the case'
+
+ocil: |-
+    To ensure the failed password attempt policy is configured correctly, run the following command:
+    <pre>$ grep pam_tally2 /etc/pam.d/common-auth</pre>
+    The output should show <tt>auth required pam_tally2.so deny=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_deny" use="legacy"/> lock_time=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_fail_interval" use="legacy"/> even_deny_root root_unlock_time=60 unlock_time=<sub idref="xccdf_org.wrlinux.content_value_var_accounts_passwords_pam_faillock_unlock_time" use="legacy"/> </tt>.
+    <pre>$ grep pam_tally2 /etc/pam.d/common-account</pre>
+    The output should show <tt>account required pam_tally2.so</tt>.
+
+platform: pam
diff --git a/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pwquality/accounts_passwords_pam_faillock_fail_interval.var b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pwquality/accounts_passwords_pam_faillock_fail_interval.var
new file mode 100644
index 0000000..7dddcfd
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/password_quality/password_quality_pwquality/accounts_passwords_pam_faillock_fail_interval.var
@@ -0,0 +1,19 @@
+documentation_complete: true
+
+title: fail_unlock_time
+
+description: 'Interval for counting failed login attempts before account lockout'
+
+type: number
+
+operator: equals
+
+interactive: false
+
+options:
+    900: 900
+    1800: 1800
+    3600: 3600
+    86400: 86400
+    604800: 604800
+    default: 900
diff --git a/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile b/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
index 572656a..7495a83 100644
--- a/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
+++ b/wrlinux1019/profiles/draft_stig_wrlinux_disa.profile
@@ -228,6 +228,7 @@ selections:
     - audit_rules_usergroup_modification
     - accounts_passwords_pam_tally2_deny
     - accounts_passwords_pam_tally2_deny_root
+    - accounts_passwords_pam_tally2_interval
     - package_samhain_installed
     - samhain_build_database
     - samhain_periodic_cron_checking
-- 
1.9.1

